core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/buffer_v1.go

```diff
@@ -75,15 +75,17 @@ type logBuffer struct {
 	// last block number seen by the buffer
 	lastBlockSeen *atomic.Int64
 	// map of upkeep id to its queue
-	queues map[string]*upkeepLogQueue
```

Comment:
Why do we need additional slice for ids? we can use the keys of `queues`

Reply:
So this is ultimately because we want to have deterministic behaviour when we are dequeuing based on iterations, while we are adding more logs to the queues. It's a little complicated but I'll try to explain it as a series of problems and solutions (with earlier problems being less relevant to the current code state):

## Problem 1
Using a map of upkeep IDs gives us a non-deterministic iteration order; in tests we want to be able to identify what upkeeps have had logs dequeued across multiple iterations, 

## Solution 1 
Use a slice of IDs to ensure that on every iteration, we dequeue the same upkeep IDs becuase the iteration order is guaranteed

## Problem 2
We should have some other deterministic way of selecting upkeep IDs for which we want to dequeue logs

## Solution 2
Use a custom upkeep selector that operates on the upkeep ID against the current iteration; this actually makes solution 1 redundant, since iteration order becomes irrelevant, because we know that the same upkeeps will be selecte for their dequeue round, regardless of iteration order. We can now remove the slice of upkeep IDs.

## Problem 3
As more logs for more upkeep IDs are added, the dequeue process, which now uses iterations to dequeue for certain upkeep IDs, will be dequeuing logs for upkeeps, that weren't present when we initially calculated the number of iterations needed to dequeue logs for all upkeeps. Now, in a single iteration, we dequeue logs for more upkeeps than intended based on the iteration calculation, which causes us to drop some logs. The upkeeps for which we dequeue logs for becomes inconsistent, since we are relying on map iteration again.

## Solution 3
Re-introduce the slice of upkeep IDs to ensure consistent iteration order. Now, as more upkeeps are added in the enqueue process, while we are dequeuing logs, we dequeue based on a consistent order, which means we ignore logs for the same (newer) upkeep IDs. We will dequeue logs for these upkeep IDs when we recalculate the number of iterations needed to dequeue for all upkeep IDs, at the end of the current set of iterations.

OR

We can update the tests to not care specifically about what upkeep IDs have had logs dequeued, just that a certain number of upkeeps have had logs dequeued. This would allow us to omit the slice of upkeep IDs, but will provide fewer insights in our tests.